README.md
# lltsv

List specified keys of LTSV (Labeled Tab Separated Values)

# Description

`lltsv` is a command line tool written in golang to list specified keys of LTSV (Labeled Tab Separated Values) text.

Example 1: 

```bash
$ echo "foo:aaa\tbar:bbb\tbaz:ccc" | lltsv -k foo,bar
foo:aaa   bar:bbb
```

The output is colorized as default when you outputs to a terminal. 
The coloring is disabled if you pipe or redirect outputs.

Example 2:

```bash
$ echo "foo:aaa\tbar:bbb\tbaz:ccc" | lltsv -k foo,bar -K
aaa       bbb
```

Eliminate labels with `-K` option.

Example 3:

```bash
$ lltsv -k foo,bar -K file*.log
```

Specify input files as arguments.

Example4:

```bash
$ lltsv -k resptime,status,uri -f 'resptime > 6' access_log
$ lltsv -k resptime,status,uri -f 'resptime > 6' -f 'uri =~ ^/foo' access_log
```

Filter output with "-f" option. Available comparing operators are:

```
  >= > == < <=  (arithmetic (float64))
  == ==* != !=* (string comparison (string))
  =~ !~ =~* !~* (regular expression (string))
```

The comparing operators terminated by __*__ behave in case-insensitive.

You can specify multiple -f options (AND condition).

Example5:

```bash
$ lltsv -k resptime,upstream_resptime,diff -e 'diff = resptime - upstream_resptime' access_log
$ lltsv -k resptime,upstream_resptime,diff_ms -e 'diff_ms = (resptime - upstream_resptime) * 1000' access_log
```

Evaluate value with "-e" option. Available operators are:

```
  + - * / (arithmetic (float64))
```

Grep period in common log format(common log or ISO 8610) "-t" option. Available operators are:

Example6:

```bash
$ lltsv -t 'localtime=01/Jun/2018:00:00:00 +0900~30/Mar/2018:23:59:59 +0900,common' access_log
$ lltsv -t 'localtime=2018-01-01T00:00:00+0900~2018-03-30T23:59:59+0900,iso8610' access_log
```

**How Useful?**

LTSV format is not `awk` friendly (I think), but `lltsv` can help it: 

```bash
$ echo -e "time:2014-08-13T14:10:10Z\tstatus:200\ntime:2014-08-13T14:10:12Z\tstatus:500" \
  | lltsv -k time,status -K | awk '$2 == 500'
2014-08-13T14:10:12Z    500
```

Useful!

## Installation

Executable binaries are available at [releases](https://github.com/sonots/lltsv/releases).

For example, for linux x86_64, 

```bash
$ wget https://github.com/sonots/lltsv/releases/download/v0.3.0/lltsv_linux_amd64 -O lltsv
$ chmod a+x lltsv
```

If you have the go runtime installed, you may use go get. 

```bash
$ go get github.com/sonots/lltsv
```

## Usage

```
$ lltsv -h
NAME:
   lltsv - List specified keys of LTSV (Labeled Tab Separated Values)

USAGE:
   lltsv [global options] command [command options] [arguments...]

VERSION:
   0.5.1

AUTHOR(S):
   sonots <sonots@gmail.com>

COMMANDS:
   help, h      Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --key, -k                                            keys to output (multiple keys separated by ,)
   --no-key, -K                                         output without keys (and without color)
   --ignore-key value, -i value                         ignored keys to output (multiple keys separated by ,)
   --filter, -f [--filter option --filter option]       filter expression to output
   --expr, -e [--expr option --expr option]             evaluate value by expression to output
   --help, -h                                           show help
   --version, -v                                        print the version
```

## ToDo

1. write tests

## Build

To build, use go get and make

```
$ go get -d github.com/sonots/lltsv
$ cd $GOPATH/src/github.com/sonots/lltsv
$ make
```

To release binaries, I use [gox](https://github.com/mitchellh/gox) and [ghr](https://github.com/tcnksm/ghr)

```
go get github.com/mitchellh/gox
gox -build-toolchain # only first time
go get github.com/tcnksm/ghr

mkdir -p pkg && cd pkg && gox ../...
ghr vX.X.X .
```

## Contribution

1. Fork (https://github.com/sonots/lltsv/fork)
2. Create a feature branch
3. Commit your changes
4. Rebase your local changes against the master branch
5. Run test suite with the go test ./... command and confirm that it passes
6. Run gofmt -s
7. Create new Pull Request

## Copyright

See [LICENSE](./LICENSE)

## Special Thanks

This is a golang fork of perl version created by [id:key_amb](http://keyamb.hatenablog.com/). Thanks!

MEMO: golang version was 5x faster than perl version


CHANGELOG.md
## 0.7.0 (2020/11/12)

Enhancements:

* Use mattn-isatty instead of andrew-d/go-termutil #28
* Introduce Go Modules #27
* Use https instead of http in comment. #24
* Add tests. #23

## 0.6.1 (2017/07/14)

Fixes:

* Show compiler version by -v. refs: #19
* feature: introduced --ignore-key. refs #20

## 0.6.0 (2017/04/11)

Fixes:

* sort keys by default. refs: #16
* build: fixed `go get` error for ghr. refs #15

## 0.5.6 (2017/03/21)

Fixes:

* feature: introduced not-equal operators for string.
* expr: re-implemented by go package.

## 0.5.5 (2016/11/24)

Fixes:

* filter: exited when invalid filter expression was given (thanks to @cubicdaiya)

## 0.5.4 (2016/10/31)

Fixes:

* Typo fixed

## 0.5.3 (2016/10/31)

Enhancements:

* Introduce case-insentive comparing operators such as `==*`, `=~*`, `!~*` (thanks to @cubicdaiya)

## 0.5.2 (2016/10/28)

Enhancements:

* Enable perfect string match by '==' (thanks to @cubicdaiya)

Changes:

* Replace codegangsta/cli with urfave/cli (thanks to @b4b4r07)

## 0.5.1 (2016/08/25)

Enhancements:

* Fix typo

## 0.5.0 (2016/08/24)

Enhancements:

* Add new option -expr, -e (thanks to @cubicdaiya)

## 0.4.1 (2015/06/13)

Enhancements:

* filter: fixed panic error when the invalid filter expression is given. (thanks to @cubicdaiya)

## 0.4.0 (2015/10/20)

Enhancements:

* Add `filter` option (thanks to @hirose31)

## 0.3.1 (2014/11/21)

Enhancements:

* Add line feed to error messages

## 0.3.0 (2014/11/21)

Enhancements:

* Read from multiple files

## 0.2.0 (2014/08/13)

Enhancements:

* Read from a file (not only from STDIN)

Changes:

* Print all fields if -k option is not specified
* Print fields in specified keys order

## 0.1.0 (2014/08/13)

* Initial Release



eval_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestEval(t *testing.T) {
	assert := assert.New(t)

	tests := []struct {
		expression string
		vars       Vars
		result     string
	}{
		{ // no space
			expression: "1+1",
			vars:       nil,
			result:     "2",
		},
		{
			expression: "1 + 1",
			vars:       nil,
			result:     "2",
		},
		{
			expression: "resptime - upstream_resptime",
			vars:       Vars{"resptime": "5", "upstream_resptime": "3"},
			result:     "2",
		},
		{
			expression: "(resptime - upstream_resptime) * 1000",
			vars:       Vars{"resptime": "5", "upstream_resptime": "3"},
			result:     "2000",
		},
	}

	for _, test := range tests {
		expr, err := parseExpr(test.expression)
		assert.Nil(err)

		ctx := &ExprContext{
			vars: test.vars,
		}
		v, err := evalExpr(expr, ctx)
		assert.Nil(err)
		assert.Equal(test.result, v.String())
	}
}


lltsv_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestFilter(t *testing.T) {
	assert := assert.New(t)

	tests := []struct {
		filter string
		lvs    map[string]string
		result bool
	}{
		// arithmetic comparison
		{
			filter: "resptime > 6",
			lvs:    map[string]string{"resptime": "10"},
			result: true,
		},
		{
			filter: "resptime >= 6",
			lvs:    map[string]string{"resptime": "6"},
			result: true,
		},
		{
			filter: "resptime == 60",
			lvs:    map[string]string{"resptime": "60"},
			result: true,
		},
		{
			filter: "resptime < 6",
			lvs:    map[string]string{"resptime": "7"},
			result: false,
		},
		{
			filter: "resptime <= 6",
			lvs:    map[string]string{"resptime": "7"},
			result: false,
		},
		// string comparison
		{
			filter: "uri == /top",
			lvs:    map[string]string{"uri": "/top"},
			result: true,
		},
		{
			filter: "uri ==* /TOP",
			lvs:    map[string]string{"uri": "/top"},
			result: true,
		},
		{
			filter: "uri != /top",
			lvs:    map[string]string{"uri": "/bottom"},
			result: true,
		},
		{
			filter: "uri !=* /top",
			lvs:    map[string]string{"uri": "/TOP"},
			result: false,
		},
		// regular expression
		{
			filter: "uri =~ ^/",
			lvs:    map[string]string{"uri": "/top"},
			result: true,
		},
		{
			filter: "uri !~ ^/",
			lvs:    map[string]string{"uri": "/top"},
			result: false,
		},
		{
			filter: "uri =~* ^/",
			lvs:    map[string]string{"uri": "/TOP"},
			result: true,
		},
		{
			filter: "uri !~* /top",
			lvs:    map[string]string{"uri": "/TOP"},
			result: false,
		},
	}

	for _, test := range tests {
		filters := getFuncFilters([]string{test.filter})
		for k, filter := range filters {
			assert.Equal(test.result, filter(test.lvs[k]))
		}
	}
}


eval.go
package main

import (
	"errors"
	"go/ast"
	"go/constant"
	"go/parser"
	"strconv"
)

// Vars is a map for LTSV's label and value.
type Vars map[string]string

// ExprRunner is an expression runner for Go code.
type ExprRunner struct {
	expr ast.Expr
}

// ExprContext is a context for ExprRunner.
type ExprContext struct {
	vars Vars
}

func parseExpr(e string) (ast.Expr, error) {
	expr, err := parser.ParseExpr(e)
	if err != nil {
		return nil, err
	}

	return expr, nil
}

func evalExpr(expr ast.Expr, ctx *ExprContext) (constant.Value, error) {
	switch e := expr.(type) {
	case *ast.BasicLit:
		return evalBasicLit(e, ctx)
	case *ast.BinaryExpr:
		return evalBinaryExpr(e, ctx)
	case *ast.Ident:
		return evalIdent(e, ctx)
	case *ast.ParenExpr:
		return evalExpr(e.X, ctx)
	}

	return constant.MakeUnknown(), errors.New("unknown expr")
}

func evalBasicLit(expr *ast.BasicLit, ctx *ExprContext) (constant.Value, error) {
	return constant.MakeFromLiteral(expr.Value, expr.Kind, 0), nil
}

func evalBinaryExpr(expr *ast.BinaryExpr, ctx *ExprContext) (constant.Value, error) {
	x, err := evalExpr(expr.X, ctx)
	if err != nil {
		return constant.MakeUnknown(), err
	}

	y, err := evalExpr(expr.Y, ctx)
	if err != nil {
		return constant.MakeUnknown(), err
	}

	return constant.BinaryOp(x, expr.Op, y), nil
}

func evalIdent(expr *ast.Ident, ctx *ExprContext) (constant.Value, error) {
	name, ok := ctx.vars[expr.Name]
	if !ok {
		return constant.MakeUnknown(), errors.New("unknown variable name")
	}

	n, err := strconv.ParseFloat(name, 64)
	if err != nil {
		return constant.MakeUnknown(), errors.New("variable type must be numeric")
	}

	return constant.MakeFloat64(n), nil
}


lltsv.go
package main

import (
	"bufio"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"

	"github.com/mattn/go-isatty"
	"github.com/mgutz/ansi"
)

type tFuncAppend func([]string, string, string) []string
type tFuncFilter func(string) bool

// Lltsv is a context for processing LTSV.
type Lltsv struct {
	keys         []string
	ignoreKeyMap map[string]struct{}
	noKey        bool
	filters      []string
	exprs        []string
	funcAppend   tFuncAppend
	funcFilters  map[string]tFuncFilter
	exprRunners  map[string]*ExprRunner
}

func newLltsv(keys []string, ignoreKeys []string, noKey bool, filters []string, exprs []string) *Lltsv {
	ignoreKeyMap := make(map[string]struct{})
	for _, key := range ignoreKeys {
		ignoreKeyMap[key] = struct{}{}
	}
	return &Lltsv{
		keys:         keys,
		ignoreKeyMap: ignoreKeyMap,
		noKey:        noKey,
		filters:      filters,
		exprs:        exprs,
		funcAppend:   getFuncAppend(noKey),
		funcFilters:  getFuncFilters(filters),
		exprRunners:  getExprRunners(exprs),
	}
}

func (lltsv *Lltsv) scanAndWrite(file *os.File) error {
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		lvs, keys := lltsv.parseLtsv(line)
		lltsv.expr(lvs)

		if lltsv.filter(lvs) {
			ltsv := lltsv.restructLtsv(lvs, keys)
			os.Stdout.WriteString(ltsv + "\n")
		}

	}
	return scanner.Err()
}

func (lltsv *Lltsv) filter(lvs map[string]string) bool {
	shouldOutput := true

	for key, funcFilter := range lltsv.funcFilters {
		if !funcFilter(lvs[key]) {
			shouldOutput = false
			break
		}
	}

	return shouldOutput
}

func (lltsv *Lltsv) expr(lvs map[string]string) {
	ctx := &ExprContext{
		vars: lvs,
	}
	for key, runner := range lltsv.exprRunners {
		v, err := evalExpr(runner.expr, ctx)
		if err == nil {
			lvs[key] = v.String()
		}
	}
}

// lvs: label and value pairs
func (lltsv *Lltsv) restructLtsv(lvs map[string]string, keys []string) string {
	// specified keys or all keys
	orders := lltsv.keys
	if len(lltsv.keys) == 0 {
		orders = keys
	}
	// make slice with enough capacity so that append does not newly create object
	// cf. https://golang.org/pkg/builtin/#append
	selected := make([]string, 0, len(orders))
	for _, label := range orders {
		if _, ok := lltsv.ignoreKeyMap[label]; ok {
			continue
		}
		value := lvs[label]
		selected = lltsv.funcAppend(selected, label, value)
	}
	return strings.Join(selected, "\t")
}

func (lltsv *Lltsv) parseLtsv(line string) (map[string]string, []string) {
	columns := strings.Split(line, "\t")
	lvs := make(map[string]string)
	keys := make([]string, 0, len(columns))
	for _, column := range columns {
		lv := strings.SplitN(column, ":", 2)
		if len(lv) < 2 {
			continue
		}
		label, value := lv[0], lv[1]
		lvs[label] = value
		keys = append(keys, label)
	}
	return lvs, keys
}

// Return function pointer to avoid `if` evaluation occurs in each iteration
func getFuncAppend(noKey bool) tFuncAppend {
	if noKey {
		return func(selected []string, label string, value string) []string {
			return append(selected, value)
		}
	}

	if isatty.IsTerminal(os.Stdout.Fd()) {
		return func(selected []string, label string, value string) []string {
			return append(selected, ansi.Color(label, "white+bh:green")+":"+ansi.Color(value, "white+bh:magenta"))
		}
	}

	// if pipe or redirect
	return func(selected []string, label string, value string) []string {
		return append(selected, label+":"+value)
	}
}

func getFuncFilters(filters []string) map[string]tFuncFilter {
	funcFilters := map[string]tFuncFilter{}
	for _, f := range filters {
		token := strings.SplitN(f, " ", 3)
		if len(token) < 3 {
			log.Fatalf("filter expression is invalid: %s\n", f)
		}
		key := token[0]
		switch token[1] {
		case ">", ">=", "<=", "<":
			r, err := strconv.ParseFloat(token[2], 64)
			if err != nil {
				log.Fatal(err)
			}

			funcFilters[key] = func(val string) bool {
				num, err := strconv.ParseFloat(val, 64)
				if err != nil {
					log.Println(err)
					return false
				}
				switch token[1] {
				case ">":
					return num > r
				case ">=":
					return num >= r
				case "<=":
					return num <= r
				case "<":
					return num < r
				default:
					return false
				}
			}
		case "==":
			funcFilters[key] = func(val string) bool {
				return val == token[2]
			}
		case "==*":
			funcFilters[key] = func(val string) bool {
				return strings.ToLower(val) == strings.ToLower(token[2])
			}
		case "!=":
			funcFilters[key] = func(val string) bool {
				return val != token[2]
			}
		case "!=*":
			funcFilters[key] = func(val string) bool {
				return strings.ToLower(val) != strings.ToLower(token[2])
			}
		case "=~", "!~", "=~*", "!~*":
			if token[1] == "=~*" || token[1] == "!~*" {
				token[2] = strings.ToLower(token[2])
			}
			re := regexp.MustCompile(token[2])
			funcFilters[key] = func(val string) bool {
				switch token[1] {
				case "=~":
					return re.MatchString(val)
				case "!~":
					return !re.MatchString(val)
				case "=~*":
					return re.MatchString(strings.ToLower(val))
				case "!~*":
					return !re.MatchString(strings.ToLower(val))
				default:
					return false
				}
			}
		}
	}
	return funcFilters
}

func getExprRunners(exprs []string) map[string]*ExprRunner {
	funcExprs := make(map[string]*ExprRunner, len(exprs))
	for _, f := range exprs {
		token := strings.SplitN(f, "=", 2)
		if len(token) != 2 {
			log.Printf("expression is invalid: %s\n", f)
			continue
		}

		expr, err := parseExpr(token[1])
		if err != nil {
			log.Printf("expression is invalid: %s\n", f)
			continue
		}

		key := strings.Trim(token[0], " ")

		funcExprs[key] = &ExprRunner{
			expr: expr,
		}
	}
	return funcExprs
}


main.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strings"

	"github.com/urfave/cli"
)

// os.Exit forcely kills process, so let me share this global variable to terminate at the last
var exitCode = 0

func main() {
	app := cli.NewApp()
	app.Name = "lltsv"
	app.Version = Version
	app.Usage = `List specified keys of LTSV (Labeled Tab Separated Values)

	Example1 $ echo "foo:aaa\tbar:bbb" | lltsv -k foo,bar
	foo:aaa   bar:bbb

	The output is colorized as default when you outputs to a terminal.
	The coloring is disabled if you pipe or redirect outputs.

	Example2 $ echo "foo:aaa\tbar:bbb" | lltsv -k foo,bar -K
	aaa       bbb

	Eliminate labels with "-K" option.

	Example3 $ lltsv -k foo,bar -K file*.log

	Specify input files as arguments.

	Example4 $ lltsv -k resptime,status,uri -f 'resptime > 6' access_log
	         $ lltsv -k resptime,status,uri -f 'resptime > 6' -f 'uri =~ ^/foo' access_log

	Filter output with "-f" option. Available comparing operators are:

    >= > == < <=  (arithmetic (float64))
    == ==* != !=* (string comparison (string))
    =~ !~ =~* !~* (regular expression (string))

        The comparing operators terminated by * behave in case-insensitive.

	You can specify multiple -f options (AND condition).

	Example5 $ lltsv -k resptime,upstream_resptime,diff -f 'diff = resptime - upstream_resptime' access_log
	         $ lltsv -k resptime,upstream_resptime,diff_ms -e 'diff_ms = (resptime - upstream_resptime) * 1000' access_log

	Evaluate value with "-e" option. Available operators are:

	  + - * / (arithmetic (float64))

	Grep period in common log format(common log or ISO 8610) "-t" option. Available operators are:

	Example6 $ lltsv -t 'localtime=01/Jun/2018:00:00:00 +0900~30/Mar/2018:23:59:59 +0900,common' access_log
	         $ lltsv -t 'localtime=2018-01-01T00:00:00+0900~2018-03-30T23:59:59+0900,iso8610' access_log

	Homepage: https://github.com/sonots/lltsv`
	app.Author = "sonots"
	app.Email = "sonots@gmail.com"
	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:  "key, k",
			Usage: "keys to output (multiple keys separated by ,)",
		},
		cli.BoolFlag{
			Name:  "no-key, K",
			Usage: "output without keys (and without color)",
		},
		cli.StringFlag{
			Name:  "ignore-key, i",
			Usage: "ignored keys to output (multiple keys separated by ,)",
		},
		cli.StringSliceFlag{
			Name:  "filter, f",
			Usage: "filter expression to output",
		},
		cli.StringSliceFlag{
			Name:  "expr, e",
			Usage: "evaluate value by expression to output",
		},
		cli.StringSliceFlag{
			Name:  "timegrep, t",
			Usage: "grep period of time in common log format or iso8601",
		},
	}
	app.Action = doMain

	cli.VersionPrinter = func(c *cli.Context) {
		fmt.Fprintf(app.Writer, `%v version %v
Compiler: %s %s
`,
			app.Name,
			app.Version,
			runtime.Compiler,
			runtime.Version())
	}

	app.Run(os.Args)
	os.Exit(exitCode)
}

func doMain(c *cli.Context) error {
	keys := make([]string, 0, 0) // slice with length 0
	if c.String("key") != "" {
		keys = strings.Split(c.String("key"), ",")
	}
	noKey := c.Bool("no-key")
	filters := c.StringSlice("filter")
	exprs := c.StringSlice("expr")
	timegrep := c.StringSlice("timegrep")

	ignoreKeys := make([]string, 0, 0)
	// If -k,--key is specified, -i,--ignore-key is ignored.
	if len(keys) == 0 && c.String("ignore-key") != "" {
		ignoreKeys = strings.Split(c.String("ignore-key"), ",")
	}

	lltsv := newLltsv(keys, ignoreKeys, noKey, filters, exprs, timegrep)

	if len(c.Args()) > 0 {
		for _, filename := range c.Args() {
			file, err := os.Open(filename)
			if err != nil {
				os.Stderr.WriteString("failed to open and read `" + filename + "`.\n")
				exitCode = 1
				return err
			}
			err = lltsv.scanAndWrite(file)
			file.Close()
			if err != nil {
				os.Stderr.WriteString("reading input errored\n")
				exitCode = 1
				return err
			}
		}
	} else {
		file := os.Stdin
		err := lltsv.scanAndWrite(file)
		file.Close()
		if err != nil {
			os.Stderr.WriteString("reading input errored\n")
			exitCode = 1
			return err
		}
	}

	return nil
}
